### PR TITLE
Make detach return an alias even under inference mode

### LINKED ISSUE #59614: inference_mode causes segfault
## To Reproduce

1. Create new, empty conda environment
2. Run ```conda install pytorch cpuonly -c pytorch-nightly```

Run
```
import torch
with torch.inference_mode():
    torch.nn.Linear(9, 64)
```
which results in a segmentation fault.

When run outside of inference_mode context the code succeeds

```
import torch
torch.nn.Linear(9, 64)
```

## Expected behavior

Don't segfault.

## Environment

$ python collect_env.py
/fsx/users/cpuhrsch/conda/envs/nightly20210607/lib/python3.9/site-packages/torch/package/_mock_zipreader.py:17: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at  /opt/conda/conda-bld/pytorch_1623049563538/work/torch/csrc/utils/tensor_numpy.cpp:68.)
  _dtype_to_storage = {data_type(0).dtype: data_type for data_type in _storages}
Collecting environment information...

CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.27

Python version: 3.9 (64-bit runtime)
Python platform: Linux-5.4.0-1041-aws-x86_64-with-glibc2.27
Is CUDA available: False
CUDA runtime version: 11.1.105
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] torch==1.10.0.dev20210607
[conda] blas                      1.0                         mkl
[conda] cpuonly                   1.0                           0    pytorch-nightly
[conda] mkl                       2021.2.0           h06a4308_296
[conda] pytorch                   1.10.0.dev20210607     py3.9_cpu_0  [cpuonly]  pytorch-nightly

## Additional context

I used git's bisect functionality and narrowed it down to https://github.com/pytorch/pytorch/pull/56017
I also confirmed the segfault on a recent (1.10.0.dev20210607-py3.8_0) on my macbook.
I can provide a backtrace if needed, but this should be easily reproduced given the brevity of the snippet, if this bug is legitimate.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @Varal7

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59633 Make detach return an alias even under inference mode**

Fixes #59614

This fix isn't 100% correct but it appears to stem the bleeding.
A better fix would be understand how to detect when function
implementations don't uphold required invariants, leading to
refcount disaster.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28962183](https://our.internmc.facebook.com/intern/diff/D28962183)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28962183).

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
this is BC-breaking, no?

**Comment by gchanan** [comment]
I guess only if AutoNonVariableType was on, which is rare.

**Comment by gchanan** [comment]
> This fix isn't 100% correct but it appears to stem the bleeding.

why isn't it correct?  Btw does this work with sparse?  I think it will, I just recall we didn't bind `alias` in python previously and `x[:]` didn't work.

> A better fix would be understand how to detect when function implementations don't uphold required invariants, leading to refcount disaster.

is it actually not legal to return non-aliases?

**Comment by ezyang** [comment]
> this is BC-breaking, no? I guess only if AutoNonVariableType was on, which is rare.

Yeah, this shouldn't actually be BC breaking, because prior to `inference_mode` it wasn't actually possible to trigger this codepath from Python, and I think in C++ we don't really exercise this. One possibility is this may cause some minor regressions on inference in production if people were calling detach.

> why isn't it correct?

I think that this specific change only improves correctness, but what I am currently missing is a good explanation of how all of these "autograd only" operations should operate when AutoNonVariableType is on. For example, we also make `detach_` a no op (not modified in this PR) and I think this is wrong, but I have to think a lot more carefully about what a good resolution in that case is. I'm also unsure about whether or not you should ever be actually exercising this code 

> Btw does this work with sparse? I think it will, I just recall we didn't bind alias in python previously and x[:] didn't work.

Nope it doesn't.

```
>>> with torch.inference_mode(): torch.sparse_coo_tensor([[1]], [2]).detach()
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: sparse tensors do not have strides
```

> is it actually not legal to return non-aliases?

The correctness conditions here are a little vague (e.g., because `contiguous` is OK with returning the same Tensor), but in general anywhere in PyTorch where we take a returned output and mutate it, the result MUST be an alias because otherwise you will mutate an input (not intended). One concrete example of this (but not applicable to the bug this PR fixes) is in autograd, where we mutate output tensors to set their `grad_fn`.

cc @robieta this is the same thing we are arguing about in #59389

**Comment by ezyang** [comment]
blegh mobile is throwing a hissy fit. Gonna do this slightly differently

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28962183).

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -2170,6 +2170,12 @@ Tensor alias(const Tensor& self) {
     return alias_with_sizes_and_strides(self, self.sizes(), self.strides());
 }
 
+Tensor detach(const Tensor& self) {
```
Comment (albanD) [inline aten/src/ATen/native/TensorShape.cpp:2173]: What is the low level implication of moving detach here without any modification of native_functions.yaml?

Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:2173]: doesn't matter. I just put it here so that it's close to alias.

Comment (albanD) [reply aten/src/ATen/native/TensorShape.cpp:2173]: Ho ok :p

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**